File: teaching/lesson2/Student.scala
class Student(var name: String, var scores:Array[Int]) {

  def averageScore: Double = {
    var sum = 0.0
    for (v <- this.scores) {
      sum += v
    }
    return sum / this.scores.length;
  }
}

def main() = {
var s0 = new Student("David L", Array(90, 86, 60, 10))
var s1 = new Student("Jovani K", Array(92, 84, 83, 12))
var s2 = new Student("Alexa O", Array(43, 23, 22, 5))

var students = List(s0, s1, s2)
students = students.sortBy(- _.averageScore)
//students.sort((e1, e2) => (e1.averageScore < e2.averageScore))

System.out.println("Ranking of students:")
for (s <- students) {
  var name = s.name
  var score = s.averageScore
  System.out.println(f"$name%s \t($score%.2f)")
}

}

main()

File: teaching/lesson2/student.go
package main

import (
  "fmt"
  "sort"
)

type Student struct {
  name string
  scores []int
}

func (s Student) score() float64 {
  val := 0.0
  for _, testScore := range s.scores {
    val += float64(testScore)
  }
  return val / float64(len(s.scores))
}

// ByAge implements sort.Interface for []Person based on
// the Age field.
type ByAge []Student

func (a ByAge) Len() int           { return len(a) }
func (a ByAge) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
func (a ByAge) Less(i, j int) bool { return a[i].score() > a[j].score() }

func main() {
  fmt.Println("Ranking of students:");
  s1 := Student{
    "David L", []int{90, 86, 60, 10}};
  s2 := Student{
    "Jovani K", []int{92, 84, 83, 12}};
  s3 := Student{
    "Alexa O", []int{43, 23, 22, 5}};
  students := []Student{s1, s2, s3}
  sort.Sort(ByAge(students))
  for _, s := range students {
    fmt.Printf("%s \t(%.2f)\n", s.name,
      s.score())
  }
}